src/PotentialEnergyEquationTerms.jl

```diff
@@ -5,11 +5,22 @@ using DocStringExtensions
 export PotentialEnergy
 
 using Oceananigans.AbstractOperations: KernelFunctionOperation
-using Oceananigans: Models.seawater_density
-using Oceananigans: Models.model_geopotential_height
+using Oceananigans.Models: seawater_density
+using Oceananigans.Models: model_geopotential_height
 using Oceananigans.Grids: Center, Face
-using Oceanostics: validate_location, validate_buoyancy
+using Oceananigans.BuoyancyModels: Buoyancy, BuoyancyTracer, SeawaterBuoyancy, LinearEquationOfState
+using Oceananigans.BuoyancyModels: buoyancy_perturbationᶜᶜᶜ
+using Oceananigans.Models: ShallowWaterModel
+using Oceanostics: validate_location
+using SeawaterPolynomials: BoussinesqEquationOfState
 
+const NoBuoyancyModel = Union{Nothing, ShallowWaterModel}
+const BuoyancyTracerModel = Buoyancy{<:BuoyancyTracer, g} where g
+const BuoyancyLinearEOSModel = Buoyancy{<:SeawaterBuoyancy{FT, <:LinearEquationOfState, T, S}} where {FT, T, S}
+const BuoyancyBoussinesqEOSModel = Buoyancy{<:SeawaterBuoyancy{FT, <:BoussinesqEquationOfState, T, S}} where {FT, T, S}
+
+linear_eos_buoyancy(model) = linear_eos_buoyancy(model.grid, model.buoyancy.model, model.tracers)
+linear_eos_buoyancy(grid, buoyancy, tracers) = KernelFunctionOperation{Center, Center, Center}(buoyancy_perturbationᶜᶜᶜ, grid, buoyancy, tracers)
```

Comment:
Here you're defining a new `KernelFunctionOperation` that wraps around `buoyancy_perturbationᶜᶜᶜ`, which then is called again in defining the `PotentialEnergy` kernel.

I think that has compilation complexity costs (when you nest too many operations sometimes the GPU can't compile it). Can we just use `buoyancy_perturbationᶜᶜᶜ` and define only one `KernelFunctionOperation`?

Reply:
Yep sure! I have also changed 

```julia
Z = model_geopotential_height(model)
```

which was also a `KernelFunctionOperation` to `Zᶜᶜᶜ`.

The only other thing on this front is that `seawater_density` returns a `KernelFunctionOperation` which creates a nested operation. If you think better this can be changed would just require a little bit of a rework.